connector/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryUtil.java

```diff
@@ -122,4 +128,45 @@ public static String friendlyTableName(TableId tableId) {
   public static void convertAndThrow(BigQueryError error) {
     throw new BigQueryException(UNKNOWN_CODE, error.getMessage(), error);
   }
+
+  /**
+   * Solving Issue #248. As the bigquery load API can handle wildcard characters, we replace (a) 10
+   * URIs ending in XXX[0-9].suffix with XXX*.suffix; (b) 10 URIs ending in XXX[0-9]*.suffix with
+   * XXX*.suffix; This reduces the total number of URIs sent to the BigQuery API
+   *
+   * @param uris the list of URIs where the data was written to
+   * @param prefixRegex the regex to catch the part before the file index.
+   * @param suffixRegex the regex to catch the part after the file index.
+   */
+  public static List<String> optimizeLoadUriList(
+      List<String> uris, String prefixRegex, String suffixRegex) {
+    Pattern pattern = Pattern.compile("(" + prefixRegex + "\\d*)\\d\\**(" + suffixRegex + ")");
+    ImmutableList.Builder<String> result = ImmutableList.builder();
+    List<String> workingList = uris;
+    Multimap<String, String> trimmedUriMap = HashMultimap.create();
+    while (!workingList.isEmpty()) {
```

Comment:
Can we introduce a helper method trimUris which would be cleaner?

while (true) {
  trimmedUris = trimUris(currentUris);
  if (trimmedUris.size() == currentUris.size()) {
     // Couldn't reduce further.
    break;
  }
  currentUri = trimmedUril;
}

Reply:
Notice that in case the collection of the mapped URIs contains less than 10 URIs then a wildcard is not being used and they are added to the result list. Moving the entire logic into the helper method would mean to process those several times in case of a large result. I took your advice for cleaning the method and moved the trimming part to another method.